suite/tests/linux/syscall_pwait.c

```diff
@@ -117,6 +126,7 @@ main(int argc, char *argv[])
 
     print("Testing epoll_pwait\n");
 
+    *ready_to_listen = true;
```

Comment:
Along the same lines: the compiler could move this store potentially below the loop (see  https://github.com/DynamoRIO/dynamorio/blob/master/core/unix/signal.c#L5864 where the compiler was moving a store to the other side of a loop) and we'd have a hang with the child never sending its signals.

Reply:
That is definitely true (the second part 'inherently racy' I agree). The fork-part I think is ok, but I am going to put in a few barriers now (not to mix up compiler and HW barriers here, the kernel has macros for both). barrier(), wmb(). I wonder if _Android_ supports the same macros?

Reply:
First, if you could explain what is being fixed by this PR's "ready_to_listen"?  As I mentioned earlier I don't understand how it's fully fixing a race?  Second, if it is a full fix, I would again vote for higher-level abstractions (e.g., threads and condvar.h) instead of raw stores and barriers, for the reasons listed above: don't roll your own synch if there's an alternative.

Reply:
It just narrows the window of the race that can happen that the child is sending the signals before the parent is waiting for them. Maybe we (I) should just wait and re-write the test in a more deterministic way (barriers or as you mentioned w/ threads, if possible)

Reply:
If my analysis above is correct, it is not possible to avoid the race using any user-mode synchronization.  The only way is to block SIGUSR2.  However, I do not see how the Android test produced the output that it did: it seems it should hang if it fails.  Repeating an earlier question: do you have a theory for how it produced that output?

Reply:
>> do you have a theory for how it produced that output?
I don't, in particular, I can't reproduce on Android. Is it a sporadic failure or does it always fail on Android? If always, maybe bionic implements this such that even though epoll_pwait unblocks USR1, the signal handler isn't called?

Reply:
The missing check_signals_pending() I mentioned above is my theory as to why the signal handler wouldn't be called -- but I don't see why it would be android-specific and I think it would be called later anyway?  I would have to review the signal code: maybe not: maybe there is some non-determinism there.

> Is it a sporadic failure or does it always fail on Android?
The CDash history on the CDash site should answer that.

Reply:
Ok, so I just re-factored, without re-writing everything using threads, but adding proper barriers and  sending the (both blocked-) signals "one-by-one". This should be deterministic/safe. It actually exposes the missing check_signals_pending() perfectly because now, the test ensures the signals are fully "pending" and when running under DR, we never call the signal handler. Is there an open bug? We can use this as a test.

Reply:
Could consider #3336 as the bug b/c it's the only theory for the missing handler calls.

Reply:
i#3336 now considered "pending signal" bug. See above, patch updated.